consensus/src/pipeline/pipeline_builder.rs

```diff
@@ -0,0 +1,714 @@
+                block.clone(),
+            ),
+            &mut abort_handles,
+        );
+        let all_fut = PipelineFutures {
+            prepare_fut,
+            execute_fut,
+            ledger_update_fut,
+            post_ledger_update_fut,
+            commit_vote_fut,
+            pre_commit_fut,
+            post_pre_commit_fut,
+            commit_ledger_fut,
+            post_commit_fut,
+        };
+        tokio::spawn(Self::monitor(
+            block.epoch(),
+            block.round(),
+            block.id(),
+            all_fut.clone(),
+        ));
+        (all_fut, tx, abort_handles)
+    }
+
+    /// Precondition: Block is inserted into block tree (all ancestors are available)
+    /// What it does: Wait for all data becomes available and verify transaction signatures
+    async fn prepare(preparer: Arc<BlockPreparer>, block: Arc<Block>) -> TaskResult<PrepareResult> {
+        let _tracker = Tracker::new("prepare", &block);
+        // the loop can only be abort by the caller
+        let input_txns = loop {
+            match preparer.prepare_block(&block).await {
+                Ok(input_txns) => break input_txns,
+                Err(e) => {
+                    warn!(
+                        "[BlockPreparer] failed to prepare block {}, retrying: {}",
+                        block.id(),
+                        e
+                    );
+                    tokio::time::sleep(Duration::from_millis(100)).await;
+                },
+            }
+        };
+        let sig_verification_start = Instant::now();
+        let sig_verified_txns: Vec<SignatureVerifiedTransaction> = SIG_VERIFY_POOL.install(|| {
+            let num_txns = input_txns.len();
+            input_txns
+                .into_par_iter()
+                .with_min_len(optimal_min_len(num_txns, 32))
+                .map(|t| Transaction::UserTransaction(t).into())
+                .collect::<Vec<_>>()
+        });
+        counters::PREPARE_BLOCK_SIG_VERIFICATION_TIME
+            .observe_duration(sig_verification_start.elapsed());
+        Ok(Arc::new(sig_verified_txns))
+    }
+
+    /// Precondition: 1. prepare finishes, 2. parent block's phase finishes 3. randomness is available
+    /// What it does: Execute all transactions in block executor
+    async fn execute(
+        prepare_phase: TaskFuture<PrepareResult>,
+        parent_block_execute_phase: TaskFuture<ExecuteResult>,
+        randomness_rx: oneshot::Receiver<Option<Randomness>>,
+        executor: Arc<dyn BlockExecutorTrait>,
+        block: Arc<Block>,
+        is_randomness_enabled: bool,
+        validator: Arc<[AccountAddress]>,
+        onchain_execution_config: BlockExecutorConfigFromOnchain,
+    ) -> TaskResult<ExecuteResult> {
+        parent_block_execute_phase.await?;
+        let user_txns = prepare_phase.await?;
+        let maybe_rand = randomness_rx
+            .await
+            .map_err(|_| anyhow!("randomness tx cancelled"))?;
```

Comment:
Two comments here:
- We should differentiate that this step failed because of the previous step failure.
- Why not move this to a precondition step? so this method only implement current step code. And, we can also wait on these futures in parallel?

Reply:
- yes, I somehow forgot to implement the error propagation part. basically for propagated error it would be something like TaskError::Propagated instead of TaskError::Internal.

- not sure what's the second question though, anything above the tracker is considered as precondition. whether we wait together or not doesn't really matter since we need all of them to continue.